rust-src/bulletproofs/src/set_non_membership_proof.rs

```diff
@@ -0,0 +1,674 @@
+    if !rhs.is_zero_point() {
+        return Err(VerificationError::InconsistentT0);
+    }
+
+    // Part 3: Verify inner product
+    // First compute helper variables g_hat, h_prime, and P_prime
+    let g_hat = v_keys.g.mul_by_scalar(&w);
+
+    let y_inv = match y.inverse() {
+        Some(inv) => inv,
+        None => return Err(VerificationError::DivisionError),
+    };
+    let y_inv_n = z_vec(y_inv, 0, n);
+    let mut minus_e_tilde = e_tilde;
+    minus_e_tilde.negate();
+
+    // get exponent for g, i.e., [z, z, ..., z]
+    let mut gexp = vec![z; n];
+
+    let mut P_prime_exps = Vec::with_capacity(2 * n + 4);
+    P_prime_exps.append(&mut gexp);
+
+    // compute exponent for h, i.e., -s, and add it to P_prime_exps
+    for si in set_vec {
+        let mut hexpi = C::Scalar::zero();
+        hexpi.sub_assign(&si);
+        P_prime_exps.push(hexpi);
+    }
+
+    // add remaining exponents
+    P_prime_exps.push(tx);
+    P_prime_exps.push(minus_e_tilde);
+    P_prime_exps.push(C::Scalar::one());
+    P_prime_exps.push(x);
+
+    // P_prime_bases starts with G, H, and Q = g_hat
+    let mut P_prime_bases = Vec::with_capacity(2 * n + 4);
+    P_prime_bases.extend(G);
+    P_prime_bases.extend(H);
+    P_prime_bases.push(g_hat);
+
+    // add remaining bases
+    P_prime_bases.push(v_keys.h);
+    P_prime_bases.push(A);
+    P_prime_bases.push(S);
+
+    // Finally verify inner product
+    let ip_verification = verify_inner_product_with_scalars(
+        transcript,
+        &y_inv_n,
+        &P_prime_bases,
+        &P_prime_exps,
+        &proof.ip_proof,
+    );
+
+    if !ip_verification {
+        return Err(VerificationError::IPVerificationError);
+    }
+
+    Ok(())
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use pairing::bls12_381::G1;
+    type SomeCurve = G1;
+
+    /// Converts the u64 set vector into a vector over the field
+    fn get_set_vector<C: Curve>(the_set: &[u64]) -> Vec<C::Scalar> {
+        the_set.iter().copied().map(C::scalar_from_u64).collect()
+    }
+
+    /// generates several values used in tests
+    fn generate_helper_values(n: usize) -> (Generators<G1>, CommitmentKey<G1>, Randomness<G1>) {
+        let rng = &mut thread_rng();
+        let gens = Generators::generate(n, rng);
+        let b = SomeCurve::generate(rng);
+        let b_tilde = SomeCurve::generate(rng);
+        let v_keys = CommitmentKey { g: b, h: b_tilde };
+        let v_rand = Randomness::generate(rng);
+
+        (gens, v_keys, v_rand)
+    }
+
+    /// Generates commitment to v given commitment key and randomness
+    fn get_v_com(
+        v: <SomeCurve as Curve>::Scalar,
+        v_keys: CommitmentKey<G1>,
+        v_rand: Randomness<G1>,
+    ) -> Commitment<G1> {
+        let v_value = Value::<SomeCurve>::new(v);
+        let v_com = v_keys.hide(&v_value, &v_rand);
+
+        v_com
+    }
+
+    #[test]
+    /// Test whether verifying an honestly generated proof works
+    fn test_smp_prove_verify() {
```

Comment:
should we call it 'test_snmp_prove_verify()' instead?

Reply:
Yes